src/wayland/virtual_keyboard/virtual_keyboard_handle.rs

```diff
@@ -78,15 +87,23 @@ where
     ) {
         match request {
             zwp_virtual_keyboard_v1::Request::Keymap { format, fd, size } => {
-                if update_keymap(data, format, fd, size as usize) {
-                    data.has_keymap.store(true, Ordering::Relaxed);
-                }
+                update_keymap(data, format, fd, size as usize);
             }
             zwp_virtual_keyboard_v1::Request::Key { time, key, state } => {
-                if !data.has_keymap.load(Ordering::Relaxed) {
-                    virtual_keyboard.post_error(NoKeymap, "`key` sent before keymap.")
-                }
+                // Ensure keymap was initialized.
+                let virtual_data = data.handle.inner.lock().unwrap();
+                let keymap = match &virtual_data.keymap {
+                    Some(keymap) => keymap,
+                    None => {
+                        virtual_keyboard.post_error(NoKeymap, "`key` sent before keymap.");
+                        return;
```

Comment:
This is technically a change in behavior, but I'm not entirely sure why the existing code didn't abort on this error to begin with. It feels like this should either be legal, or the request should be ignored, so this change seems like an improvement to me.

Reply:
That's fine.